### PR TITLE
Viacoin: add support

### PR DESCRIPTION
**Pull request by romanornr**
Required NBXplorer pull request: https://github.com/dgarage/NBXplorer/pull/61

### CODE REVIEW

File: BTCPayServer/BTCPayNetworkProvider.Viacoin.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+            Add(new BTCPayNetwork()
+            {
+                CryptoCode = nbxplorerNetwork.CryptoCode,
+                BlockExplorerLink = NetworkType == NetworkType.Mainnet ? "https://explorer.viacoin.org/tx/{0}" : "https://explorer.viacoin.org/tx/{0}",
+                NBitcoinNetwork = nbxplorerNetwork.NBitcoinNetwork,
+                NBXplorerNetwork = nbxplorerNetwork,
+                UriScheme = "viacoin",
+                DefaultRateRules = new[] 
+                {
+                                "VIA_BTC = bittrex(VIA_BTC)"
```
Comment (NicolasDorier) [inline BTCPayServer/BTCPayNetworkProvider.Viacoin.cs]: You need to add `VIA_X = VIA_BTC * BTC_X` as well so `VIA_USD` can be solved.

**Comment by NicolasDorier** [comment]
will merge once the test for NBXplorer and NBitcoin pass. It seems it does not now.

**Comment by romanornr** [comment]
NBitcoin & NBXplorer passed. 

I just tested BtcPay with viacoin as used chain.
I think it works fine now.

screenshot from 2018-07-13 14-53-14

screenshot from 2018-07-13 14-54-04

screenshot from 2018-07-13 14-54-17

**Comment by romanornr** [comment]
Also maybe this helps. 
https://github.com/viacoin/docker/blob/master/docker-viacoin

**Comment by NicolasDorier** [comment]
@romanornr it seems very good.
Can you host your dockerfile on docker-hub so I can reference it in the one click deploy?

**Comment by romanornr** [comment]
Yes I can. 

Small "silly" question. Does BtcPay really require to have txindex=1 enabled ?
I did my test without txindex.

**Comment by romanornr** [comment]
@NicolasDorier https://hub.docker.com/r/romanornr/docker-viacoin/

**Comment by NicolasDorier** [comment]
Can you tag the image with 0.15.1 ?

**Comment by NicolasDorier** [comment]
Also, I need you to run `git update-index --chmod=+x docker-entrypoint.sh` and create a new image.

**Comment by romanornr** [comment]
I think it should be like this ?

**Comment by NicolasDorier** [comment]
for the tag yet, but I still need `git update-index --chmod=+x docker-entrypoint.sh`

**Comment by NicolasDorier** [comment]
and you need to build image again

**Comment by romanornr** [comment]
I did ```git update-index --chmod=+x docker-entrypoint.sh```
and ```git add -A```

(nothing to push)

and I rebuilded the image.

**Comment by NicolasDorier** [comment]
I can't start your image.

```
Recreating d87593368888_btcpayserver_viacoind ... error

ERROR: for d87593368888_btcpayserver_viacoind  Cannot start service viacoind: OCI runtime create failed: container_linux.go:348: starting container process caused "exec: \"/entrypoint.sh\": permission denied": unknown
ERROR: for viacoind  Cannot start service viacoind: OCI runtime create failed: container_linux.go:348: starting container process caused "exec: \"/entrypoint.sh\": permission denied": unknown
ERROR: Encountered errors while bringing up the project.
```

Try to `chmod +x docker-entrypoint.sh` as well and build the image please.

**Comment by NicolasDorier** [comment]
@romanornr I tried to fix your dockerfile on https://github.com/NicolasDorier/docker-bitcoin/tree/master/viacoin/0.15.1

There was bugs inside: 

* No chmod+x on the entrypoint
* Unzipping at the wrong place

Despite this, it still fail. I don't have time to fix it, so I let you test it and wake be up if that work. But please test it before.

```
btcpayserver_viacoind |
btcpayserver_viacoind |
btcpayserver_viacoind | ************************
btcpayserver_viacoind | EXCEPTION: N5boost10filesystem16filesystem_errorE
btcpayserver_viacoind | boost::filesystem::create_directory: File exists: "/home/viacoin/.viacoin"
btcpayserver_viacoind | viacoin in AppInit()
btcpayserver_viacoind |
btcpayserver_viacoind | viacoind: chainparamsbase.cpp:69: const CBaseChainParams& BaseParams(): Assertion `globalChainBaseParams' failed.
btcpayserver_viacoind exited with code 139
```

**Comment by chainfailure** [comment]
Can you make sure you're not reusing cached objects? With the `chmod` fix applied it just syncs properly when trying to repro here.

**Comment by NicolasDorier** [comment]
image

**Comment by NicolasDorier** [comment]
With the chmod, and BTCPay (outside of btcpay it does not crash):

```
version: "3"
services:
  btcpayserver:
    restart: unless-stopped
    image: nicolasdorier/btcpayserver:1.0.2.48
    expose:
    - "49392"
    environment:
      BTCPAY_POSTGRES: User ID=postgres;Host=postgres;Port=5432;Database=btcpayserver${NBITCOIN_NETWORK:-regtest}
      BTCPAY_NETWORK: ${NBITCOIN_NETWORK:-regtest}
      BTCPAY_BIND: 0.0.0.0:49392
      BTCPAY_EXTERNALURL: ${BTCPAY_PROTOCOL:-https}://${BTCPAY_HOST}/
      BTCPAY_ROOTPATH: ${BTCPAY_ROOTPATH:-/}
      VIRTUAL_NETWORK: nginx-proxy
      VIRTUAL_PORT: 49392
      VIRTUAL_HOST: ${BTCPAY_HOST}
      SSL_POLICY: Mozilla-Modern
      LETSENCRYPT_HOST: ${BTCPAY_HOST}
      LETSENCRYPT_EMAIL: ${LETSENCRYPT_EMAIL:-<no value>}
      BTCPAY_CHAINS: "via"
      BTCPAY_VIAEXPLORERURL: http://nbxplorer:32838/
    links:
    - nbxplorer
    - postgres
    volumes:
    - "btcpay_datadir:/datadir"
    - "nbxplorer_datadir:/root/.nbxplorer"
    ports:
    - "80:49392"
  nbxplorer:
    restart: unless-stopped
    image: nicolasdorier/nbxplorer:1.0.2.12
    expose:
    - "32838"
    environment:
      NBXPLORER_NETWORK: ${NBITCOIN_NETWORK:-regtest}
      NBXPLORER_BIND: 0.0.0.0:32838
      NBXPLORER_CHAINS: "via"
      NBXPLORER_VIARPCURL: http://viacoind:43782/
      NBXPLORER_VIANODEENDPOINT: viacoind:39388
    volumes:
    - "nbxplorer_datadir:/datadir"
    - "viacoin_datadir:/root/.viacoin"
    links:
    - viacoind
  postgres:
    restart: unless-stopped
    image: postgres:9.6.5
    volumes:
    - "postgres_datadir:/var/lib/postgresql/data"
  viacoind:
    restart: unless-stopped
    container_name: btcpayserver_viacoind
    image: nicolasdorier/docker-viacoin:0.15.1
    environment:
      VIACOIN_DATA: |
        rpcport=43782
        ${NBITCOIN_NETWORK:-regtest}=1
        port=39388
        whitelist=0.0.0.0/0
    expose:
    - "43782"
    - "39388"
    volumes:
    - "viacoin_datadir:/data"
volumes:
  postgres_datadir: 
  btcpay_datadir: 
  nbxplorer_datadir: 
  viacoin_datadir: 
```

I get the error I said.

**Comment by chainfailure** [comment]
Ah yes, able to reproduce with the compose file.

```
FROM debian:stretch-slim

RUN groupadd -r viacoin && useradd -r -m -g viacoin viacoin

RUN set -ex \
	&& apt-get update \
	&& apt-get install -qq --no-install-recommends ca-certificates dirmngr gosu gpg wget \
	&& rm -rf /var/lib/apt/lists/*

ENV VIACOIN_VERSION 0.15.1
ENV VIACOIN_URL https://github.com/viacoin/viacoin/releases/download/v0.15.1/viacoin-0.15.1-x86_64-linux-gnu.tar.gz
ENV VIACOIN_SHA256 673bfd17194ca4fe8408450e1871447d461ce26925e71ea55eebd89c379f5775

# install Viacoin binaries
RUN set -ex \
	&& cd /tmp \
	&& wget -qO viacoin.tar.gz "$VIACOIN_URL" \
	&& echo "$VIACOIN_SHA256 viacoin.tar.gz" | sha256sum -c - \
	&& tar -xzvf viacoin.tar.gz -C /usr/local --strip-components=1 --exclude=*-qt \
	&& rm -rf /tmp/*

# create data directory
# To speed up sync process: https://bootstrap.viacoin.org/
ENV VIACOIN_DATA /data
RUN mkdir "$VIACOIN_DATA" \
	&& chown -R viacoin:viacoin "$VIACOIN_DATA" \
	&& ln -sfn "$VIACOIN_DATA" /home/viacoin/.viacoin \
	&& chown -h viacoin:viacoin /home/viacoin/.viacoin
VOLUME /data

COPY docker-entrypoint.sh /entrypoint.sh
ENTRYPOINT ["/entrypoint.sh"]

EXPOSE 5222 5223 25222 25223 25222 25223
CMD ["viacoind", "--datadir=/data"]
```

Seems to work both with and without the compose context, can you confirm?

**Pull request merged**